piccolo_api/audit_logs/tables.py

```diff
@@ -0,0 +1,118 @@
+from piccolo.table import Table
+
+
+class AuditLog(Table):
+    class ActionType(str, Enum):
+        """An enumeration of AuditLog table actions type."""
+
+        CREATING = "CREATING"
+        UPDATING = "UPDATING"
+        DELETING = "DELETING"
+
+    action_time = Timestamp()
+    action_type = Varchar(choices=ActionType)
+    action_user = Varchar()
+    change_message = Text()
+    changes_in_row = JSON()
+
+    @classmethod
+    async def record_save_action(
+        cls,
+        table: t.Type[Table],
+        user_id: int,
+        new_row_id=t.Union[str, uuid.UUID, int],
+    ):
+        """
+        A method for tracking creating record actions.
+
+        :param table:
+            A table for which we monitor activities.
+        :param user_id:
+            The ``primary key`` of authenticated user.
+        """
+        result = cls(
+            action_type=cls.ActionType.CREATING,
+            action_user=cls.get_user_username(user_id),
+            change_message=f"User {cls.get_user_username(user_id)} "
+            f"create row {new_row_id} in {table._meta.tablename.title()} "
+            f"table",
+        )
+        await result.save().run()
+
+    @classmethod
+    async def record_patch_action(
+        cls,
+        table: t.Type[Table],
+        row_id: t.Union[str, uuid.UUID, int],
+        user_id: int,
+        changes_in_row: t.Dict[str, t.Any],
+    ):
+        """
+        A method for tracking updating record actions.
+
+        :param table:
+            A table for which we monitor activities.
+        :param row_id:
+            The ``primary key`` of the table for which we
+            monitor activities.
+        :param user_id:
+            The ``primary key`` of authenticated user.
+        :param changes_in_row:
+            JSON with all changed columns in the existing row.
+        """
+        result = cls(
+            action_type=cls.ActionType.UPDATING,
+            action_user=cls.get_user_username(user_id),
+            change_message=f"User {cls.get_user_username(user_id)} update row "
+            f"{row_id} in {table._meta.tablename.title()} table",
+            changes_in_row=changes_in_row,
+        )
+        await result.save().run()
+
+    @classmethod
+    async def record_delete_action(
+        cls,
+        table: t.Type[Table],
+        row_id: t.Union[str, uuid.UUID, int],
+        user_id: int,
+    ):
+        """
+        A method for tracking deletion record actions.
+
+        :param table:
+            A table for which we monitor activities.
+        :param row_id:
+            The ``primary key`` of the table for which we
+            monitor activities.
+        :param user_id:
+            The ``primary key`` of authenticated user.
+        """
+        result = cls(
+            action_type=cls.ActionType.DELETING,
+            action_user=cls.get_user_username(user_id),
+            change_message=f"User {cls.get_user_username(user_id)} delete row "
```

Comment:
I'd be tempted to get rid of the change message all together.

One option is just storing everything in JSON.

For delete, the json can be:

```
// Maps to a list, so we can also use it for a bulk delete endpoint.
{
    "deleted_ids": [1, 2]
}
```

Likewise, for the save, it could be:

```
// Maps to a list, so we can also use it for a bulk creation endpoint.
{
    "created_ids": [1, 2]
}
```

Or we have a field called `effected_rows` which is an integer or array of integers.

It's just about making it easy to query - imagine you wanted to know who deleted row 42. You would have to something like:

```python
await AuditLog.select().where(
    AuditLog.action_type == AuditLog.ActionType.DELETING,
    AuditLog.change_message.like('%42%')
)
```

It's quite imprecise, ad would be better to just do:

```python
await AuditLog.select(AuditLog.action_user).where(
    AuditLog.action_type == AuditLog.ActionType.DELETING,
    AuditLog.effected_rows.any(42)
)
```

Reply:
> I'd be tempted to get rid of the change message all together.
> 
> One option is just storing everything in JSON.
> 
> For delete, the json can be:
> 
> ```
> // Maps to a list, so we can also use it for a bulk delete endpoint.
> {
>     "deleted_ids": [1, 2]
> }
> ```
> 
> Likewise, for the save, it could be:
> 
> ```
> // Maps to a list, so we can also use it for a bulk creation endpoint.
> {
>     "created_ids": [1, 2]
> }
> ```

But all actions are for endpoints that work with a single record. We don't have the `bulk_delete` option yet ([PR here](https://github.com/piccolo-orm/piccolo_api/pull/145)) and we don't have `bulk_creation` at all. Audit logs would be primarily used in Piccolo Admin and there we delete multiple records by looping through the [delete_single](https://github.com/piccolo-orm/piccolo_admin/blob/2fbdec3a63c919bc4c140b75ec5469b4921e78d0/admin_ui/src/views/RowListing.vue#L417) endpoint.

> Or we have a field called `effected_rows` which is an integer or array of integers.

That would be possible if we have only `primary_key` type `Serial`. But we also have `primary_key` type `string` and `UUID`, and we can't use `Integer` column for that (which will be great for precise filtering). We need to use `Varchar` and query would be very similar for `effected_row` and `change_message` using `like`.

> It's just about making it easy to query - imagine you wanted to know who deleted row 42. You would have to something like:
> 
> ```python
> await AuditLog.select().where(
>     AuditLog.action_type == AuditLog.ActionType.DELETING,
>     AuditLog.change_message.like('%42%')
> )
> ```
> 
> It's quite imprecise, ad would be better to just do:
> 
> ```python
> await AuditLog.select(AuditLog.action_user).where(
>     AuditLog.action_type == AuditLog.ActionType.DELETING,
>     AuditLog.effected_rows.any(42)
> )
> ```
Here is screenshot with new columns.

![changed_columns](https://user-images.githubusercontent.com/30960668/176098525-8db82366-f53e-46f3-96d6-15dfdf993b45.png)

I don't think we should remove the `change_message` column because it nicely describes the action the user has taken. Filtering is maybe imprecise, but we can't use the Integer column for `effected_row`.

> Similarly, table name should probably be a column too, to make it easier to filter out the changes on a certain table. I'd be tempted to add `table_name` and `table_class` - again, just to make it easy to filter. What do you think?

I added `table_name`, but I don't understand what the `table_class` column would be for and what it should go into that column?

Reply:
Good points.

> But all actions are for endpoints that work with a single record. We don't have the bulk_delete option yet (https://github.com/piccolo-orm/piccolo_api/pull/145) and we don't have bulk_creation at all. Audit logs would be primarily used in Piccolo Admin and there we delete multiple records by looping through the [delete_single](https://github.com/piccolo-orm/piccolo_admin/blob/2fbdec3a63c919bc4c140b75ec5469b4921e78d0/admin_ui/src/views/RowListing.vue#L417) endpoint.

You're right, there's no bulk action endpoints at the moment. But I was just thinking about the future. Imagine someone deletes 1000 rows - we would then have to insert 1000 rows into the audit table.

> That would be possible if we have only primary_key type Serial. But we also have primary_key type string and UUID, and we can't use Integer column for that (which will be great for precise filtering). We need to use Varchar and query would be very similar for effected_row and change_message using like.

Good point - the primary key might not be an integer. Storing them all as strings is OK.

> I don't think we should remove the change_message column because it nicely describes the action the user has taken. Filtering is maybe imprecise, but we can't use the Integer column for effected_row.

Maybe `change_message` could be achieved using `get_readable` instead.

> I added table_name, but I don't understand what the table_class column would be for and what it should go into that column?

I was just thinking that if there's a table class called `MyTable`, and it refers to the database table `my_table`, maybe we would want to store them both. But it's not that important. Just need to decide which one is best to store - `my_table` or `MyTable`.

Reply:
> You're right, there's no bulk action endpoints at the moment. But I was just thinking about the future. Imagine someone deletes 1000 rows - we would then have to insert 1000 rows into the audit table.

I agree, but we don’t currently have the classic bulk delete in Piccolo Admin (we loop through `delete_single` endpoint) and audit logs would primarily be used in Piccolo Admin (page size limit is 100 per page, which is not so bad).
Once we have the proper `bulk_delete`, we can change the delete action for the audit logs to something like this

```python
# in piccolo_api.crud.endpoints.py
@apply_validators
async def delete_bulk(
self, request: Request, params: t.Optional[t.Dict[str, t.Any]] = None
) -> Response:
    """
    Bulk deletes rows - query parameters are used for filtering.
    """
    params = self._clean_data(params) if params else {}
    split_params = self._split_params(params)
    split_params_ids = split_params.ids.split(",")
  
    try:
        query: t.Union[
            Select, Count, Objects, Delete
        ] = self.table.delete()
        try:
            # Serial or UUID primary keys enabled in query params
            value_type = self.table._meta.primary_key.value_type
            ids = [value_type(item) for item in split_params_ids]
            query_ids = query.where(
                self.table._meta.primary_key.is_in(ids)
            )
            query = self._apply_filters(query_ids, split_params)
            # this is not tested but it should be work
            # in ids list can be one or more items 
            deleted_rows_map = {"deleted_pks": ids}
            try:
                if self.audit_log_table:
                    await self.audit_log_table.record_bulk_delete_action(
                        self.table,
                        deleted_rows_pk=deleted_rows_map,
                        user_id=request.user.user_id,
                    )
            except Exception as exception:
                logger.log(msg=f"{exception}", level=logging.WARNING)
        except ValueError:
            query = self._apply_filters(query, split_params)
    except MalformedQuery as exception:
         return Response(str(exception), status_code=400)

    await query.run()
    return Response(status_code=204)

# in piccolo_api.audit_logs.tables.py

class AuditLog(Table):
    ...
    deleted_rows = JSON()
    ...
    @classmethod
    async def record_delete_action(
        cls,
        table: t.Type[Table],
        deleted_rows_pk: t.Dict[str, t.Any],
        user_id: int,
    ):
        result = cls(
            action_type=cls.ActionType.deleting,
            action_user=cls.get_user_username(user_id),
            deleted_rows=deleted_rows_pk,
        )
        await result.save().run()

```
But for now we don't have that option because we're looping through the `delete_single` endpoint and we can't store multiple IDs in a row like this
```vue
// Maps to a list, so we can also use it for a bulk delete endpoint.
{
    "deleted_ids": [1, 2]
}
```
 We should stick to what we have at the moment, and later we can always change (when we have `bulk_delete` we will have to make changes also in Piccolo Admin).

> I was just thinking that if there's a table class called `MyTable`, and it refers to the database table `my_table`, maybe we would want to store them both. But it's not that important. Just need to decide which one is best to store - `my_table` or `MyTable`.

I think `table_name` is a better option because it's `Varchar` and we can use the filter sidebar to search and filter records.